### PR TITLE
Check for overflow in IncrementAccountCounter().

### LINKED ISSUE #5487: Check for overflow of the wallet account counter
This overflow is actually reachable because it's only at 2<sup>31</sup>.

See the TODO I added here: https://github.com/zcash/zcash/pull/5475#discussion_r787216107

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes #5487.

### CODE REVIEW

File: src/wallet/walletdb.h [hunk 1]
```diff
@@ -103,9 +104,16 @@ class CHDChain
...
     }
 
-    void IncrementAccountCounter() {
-        // TODO: We should check for overflow somewhere and handle it.
-        accountCounter += 1;
+    /** Increments the account counter by 1 and returns it. Returns std::nullopt
+     *  if the increment operation would cause an overflow. */
+    std::optional<uint32_t> IncrementAccountCounter() {
+        auto newAccountCounter = accountCounter + 1;
+        if (newAccountCounter > (HARDENED_KEY_LIMIT - 1)) {
```
Comment (str4d) [inline src/wallet/walletdb.h:111]: We use `HARDENED_KEY_LIMIT - 1` as the "legacy account" for deriving legacy transparent and Sapling keys, so it's always in use. The last valid value for `accountCounter` is therefore `ZCASH_LEGACY_ACCOUNT - 1`.
```suggestion
        if (newAccountCounter >= ZCASH_LEGACY_ACCOUNT) {
```

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -522,7 +522,11 @@ std::pair<UnifiedFullViewingKey, libzcash::AccountId> CWallet::GenerateNewUnifie
     while (true) {
         auto accountId = hdChain.GetAccountCounter();
         auto generated = GenerateUnifiedSpendingKeyForAccount(accountId);
-        hdChain.IncrementAccountCounter();
+        auto account = hdChain.IncrementAccountCounter();
+        if (!account.has_value()) {
+            throw std::runtime_error(
+                    "CWallet::GenerateNewUnifiedSpendingKey(): Account counter overflowed (2^31 - 1).");
```
Comment (str4d) [inline src/wallet/wallet.cpp:528]:
```suggestion
                    "CWallet::GenerateNewUnifiedSpendingKey(): Already generated the maxiumum number of accounts (2^31 - 2) for this wallet's mnemonic phrase. Congratulations, you need to create a new wallet!");
```

**Review by str4d — CHANGES_REQUESTED** [review]
utACK after comments addressed.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (nuttycom) [reply src/wallet/wallet.cpp:528]: :laughing:

**Review by nuttycom — DISMISSED** [review]
utACK

**Review by str4d — DISMISSED** [review]
utACK

File: src/wallet/walletdb.h [hunk 3]
```diff
@@ -103,9 +104,16 @@ class CHDChain
...
     }
 
-    void IncrementAccountCounter() {
-        // TODO: We should check for overflow somewhere and handle it.
-        accountCounter += 1;
+    /** Increments the account counter by 1 and returns it. Returns std::nullopt
+     *  if the increment operation would cause an overflow. */
+    std::optional<uint32_t> IncrementAccountCounter() {
+        auto newAccountCounter = accountCounter + 1;
+        if (newAccountCounter >= ZCASH_LEGACY_ACCOUNT) {
```
Comment (str4d) [inline src/wallet/walletdb.h:111]:
```suggestion
        if (newAccountCounter >= libzcash::ZCASH_LEGACY_ACCOUNT) {
```

**Review by str4d — APPROVED** [review]
re-utACK 1fb153a333929c249e28b81cae4cad91fc6c1768

**Review by nuttycom — APPROVED** [review]
re-utACK https://github.com/zcash/zcash/pull/5849/commits/1fb153a333929c249e28b81cae4cad91fc6c1768

**Review by daira — APPROVED** [review]
utACK

**Comment by mdr0id** [comment]
TestACK in CI; only fail was 1 worker in rpc:
```
feature_zip239.py:
Initializing test directory /tmp/testk6y9hopy/150
MiniNode: Connecting to Bitcoin Node IP # 127.0.0.1:12200 using version 170015
MiniNode: Connecting to Bitcoin Node IP # 127.0.0.1:12200 using version 170015
MiniNode: Connecting to Bitcoin Node IP # 127.0.0.1:12200 using version 170015
MiniNode: Connecting to Bitcoin Node IP # 127.0.0.1:12200 using version 170015
waiting for async operation opid-8054a22e-5dee-4177-8f18-fd8707fbdd54
waiting for async operation opid-044bf766-7b01-45ff-a33e-b00e7701b6bd
Assertion failed: Should have received pong
Stopping nodes
Cleaning up
Failed

stderr:
  File "/workspace/source/qa/rpc-tests/test_framework/test_framework.py", line 146, in main
    self.run_test()
  File "/workspace/source/qa/rpc-tests/feature_zip239.py", line 229, in run_test
    self.verify_inv(testnode, [
  File "/workspace/source/qa/rpc-tests/feature_zip239.py", line 72, in verify_inv
    testnode.sync_with_ping(waiting_for=lambda x: x.last_inv)
  File "/workspace/source/qa/rpc-tests/tx_expiry_helper.py", line 81, in sync_with_ping
    fail("Should have received pong")
  File "/workspace/source/qa/rpc-tests/test_framework/util.py", line 599, in fail
    raise AssertionError(message)

Pass: False, Duration: 135 s
```

**Pull request merged**